### PR TITLE
Jihwan/facilities

### PR DESCRIPTION
**Pull request by jhseung**
A general overview of the facilities in the Gym details page. Some questionable coding choices –– as this was a nested tableview that was nested within another tableview. While I wanted to make this code as reusable as possible, that would become quite messy, as the number of nested tableviews would also be dependent on the data. 

Our current goal is to simply display facilities. (Which is pretty static data, not stored in the backend database, and there is no good way for backend to scrape it either –– which meant that the data had to be hardcoded for current version's purposes). 

This feature should become more polished in the future when we find a structured way to format and retrieve the data.

### CODE REVIEW

File: Fitness.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -750,11 +762,11 @@
 				ASSETCATALOG_COMPILER_APPICON_NAME = AppIcon;
 				CODE_SIGN_IDENTITY = "iPhone Developer";
 				CODE_SIGN_STYLE = Automatic;
-				DEVELOPMENT_TEAM = "";
+				DEVELOPMENT_TEAM = A5GB6MS7YX;
 				INFOPLIST_FILE = Fitness/Info.plist;
 				LD_RUNPATH_SEARCH_PATHS = "$(inherited) @executable_path/Frameworks";
 				OTHER_SWIFT_FLAGS = "$(inherited) -DDEBUG";
-				PRODUCT_BUNDLE_IDENTIFIER = com.cornellappdev.uplift.ios;
+				PRODUCT_BUNDLE_IDENTIFIER = com.cornellappdev.uplift.ioss;
```
Comment (jmf373) [inline Fitness.xcodeproj/project.pbxproj]: Don't forget to change this back!

File: Fitness/Controllers/GymDetailViewController.swift [hunk 2]
```diff
@@ -75,13 +83,13 @@ class GymDetailViewController: UIViewController {
         super.viewDidLoad()
         view.backgroundColor = .white
         
+        print (gym.id)
```
Comment (jmf373) [inline Fitness/Controllers/GymDetailViewController.swift]: remove prints

File: Fitness/Views/GymFacilitiesView.swift [hunk 3]
```diff
@@ -0,0 +1,483 @@
...
+    ]
+    var facilityName: FacilityName!
+    var facilityData: [FacilityData]!
+    var facilityTableViews: [UITableView] = []
+    
+    var bowlingTableView: UITableView?
+    var poolTableView: UITableView?
+    var basketballTableView: UITableView?
+    var basketballSeparator: UIView?
+    var poolSeparator: UIView?
```
Comment (jmf373) [inline Fitness/Views/GymFacilitiesView.swift]: alphabetize

File: Fitness/Views/GymFacilitiesView.swift [hunk 4]
```diff
@@ -0,0 +1,483 @@
...
+    
+    func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
+        guard let unwrappedFacilityName = self.facilityName else {
+            return 0
+        }
+        
+        switch unwrappedFacilityName {
+        case .helenNewman:
+            if tableView == basketballTableView! {
+                return facilityData[0].isDropped ? 36 : 36
```
Comment (jmf373) [inline Fitness/Views/GymFacilitiesView.swift]: Any reason not to just `return 36` here or will we have varying header sizes in the future?

File: Fitness/Views/GymFacilitiesView.swift [hunk 5]
```diff
@@ -0,0 +1,483 @@
...
+                basketballTableView!.beginUpdates()
+                var modifiedIndices: [IndexPath] = []
+                for i in 0..<7 {
+                    modifiedIndices.append(IndexPath(row: i, section: 0))
+                }
+                
+                if (facilityData[0].isDropped) {
+                    facilityData[0].isDropped = false
+                    basketballTableView!.deleteRows(at: modifiedIndices, with: .fade)
+                    (basketballTableView!.headerView(forSection: 0) as! FacilityHoursHeaderView).downArrow.image = .none
```
Comment (jmf373) [inline Fitness/Views/GymFacilitiesView.swift]: Is it more efficient to set the images initially and just toggle the `isHidden` attribute?

File: Fitness/Views/Cells/FacilityHoursCell.swift [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+            make.leading.equalToSuperview().offset(22)
+            make.height.equalTo(15)
+            make.centerY.equalToSuperview()
+        }
+        
+        hoursScrollView.snp.updateConstraints {make in
+            make.leading.equalTo(dayLabel.snp.trailing).offset(18)
+            make.trailing.equalToSuperview().offset(-18)
+            make.centerY.equalToSuperview()
+            make.height.equalTo(19)
```
Comment (jmf373) [inline Fitness/Views/Cells/FacilityHoursCell.swift]: This should be a little bigger to match `hoursLabel.intrinsicContentSize.height`-- right now its a touch smaller and so we can scroll vertically in the cells which seems weird

File: Fitness/Controllers/GymDetailViewController.swift [hunk 7]
```diff
@@ -17,6 +17,12 @@ struct HoursData {
     var isDropped: Bool!
 }
 
+struct FacilityData {
+    var facility: String!
```
Comment (drewsdunne) [inline Fitness/Controllers/GymDetailViewController.swift]: Do these need to be var? Can each of these be `let name: Type` without the `!` either?

File: Fitness/Controllers/GymDetailViewController.swift [hunk 8]
```diff
@@ -375,21 +387,15 @@ class GymDetailViewController: UIViewController {
...
-                } else {
-                    make.top.equalTo(facilitiesLabelArray[i-1].snp.bottom)
-                }
-                make.centerX.width.equalToSuperview()
-                make.height.equalTo(20)
-            }
+        
+        facilityTableView.snp.updateConstraints {make in
+            make.width.centerX.equalToSuperview()
+            make.top.equalTo(facilitiesTitleLabel.snp.bottom).offset(20)
```
Comment (drewsdunne) [inline Fitness/Controllers/GymDetailViewController.swift]: Can we pull this constant out according to bible?

File: Fitness/Views/Cells/FacilityHoursCell.swift [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+    }
+    
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    // MARK: - CONSTRAINTS
+    func setupConstraints() {
+        dayLabel.snp.updateConstraints {make in
+            make.leading.equalToSuperview().offset(22)
```
Comment (drewsdunne) [inline Fitness/Views/Cells/FacilityHoursCell.swift]: Can we bring all these constants up to the top of the function according to bible?

File: Fitness/Views/GymFacilitiesView.swift [hunk 10]
```diff
@@ -0,0 +1,483 @@
...
+
+class FacilitiesGestureRecognizer: UITapGestureRecognizer {
+    var title = String()
+}
+
+class GymFacilitiesTableView: UIView, UITableViewDataSource, UITableViewDelegate {
+    
+    let helenNewmanFacilities: [FacilityData] = [
+        FacilityData(facility: "BASKETBALL COURT", isDropped: false, data: [
+            ("10 AM - 12 AM", "MON"),
```
Comment (drewsdunne) [inline Fitness/Views/GymFacilitiesView.swift]: When will this data be from backend?

File: Fitness/Views/GymFacilitiesView.swift [hunk 11]
```diff
@@ -0,0 +1,483 @@
...
+    }
+    
+    private func setupViews(facilityName: FacilityName) {
+        guard let unwrappedFacilityName = self.facilityName else {
+            return
+        }
+        
+        switch unwrappedFacilityName {
+        case .helenNewman:
+            bowlingTableView = UITableView(frame: .zero, style: .grouped)
```
Comment (drewsdunne) [inline Fitness/Views/GymFacilitiesView.swift]: How come we need so many table views in one view? Would sections not be enough with different cell classes?

File: Fitness/Views/GymFacilitiesView.swift [hunk 12]
```diff
@@ -0,0 +1,483 @@
...
+        guard let unwrappedFacilityName = self.facilityName else {
+            return
+        }
+        
+        switch unwrappedFacilityName {
+        case .helenNewman:
+            basketballTableView!.snp.updateConstraints {make in
+                make.width.centerX.top.equalToSuperview()
+                if (facilityData[0].isDropped) {
+                    make.height.equalTo(260)
```
Comment (drewsdunne) [inline Fitness/Views/GymFacilitiesView.swift]: Constants again. 260 especially as it's repeated throughout this function.

File: Fitness/Views/GymFacilitiesView.swift [hunk 13]
```diff
@@ -0,0 +1,483 @@
...
+        guard let unwrappedFacilityName = self.facilityName else {
+            return nil
+        }
+        
+        switch unwrappedFacilityName {
+        case .helenNewman:
+            let header = tableView.dequeueReusableHeaderFooterView(withIdentifier: FacilityHoursHeaderView.identifier) as! FacilityHoursHeaderView
+            var gesture: FacilitiesGestureRecognizer!
+            
+            if tableView == basketballTableView {
```
Comment (drewsdunne) [inline Fitness/Views/GymFacilitiesView.swift]: Switching on the table view seems like an odd way to do this, is there a reason this way was chosen over another?

File: Fitness/Views/Headers/FacilityHoursHeaderView.swift [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+    
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    // MARK: - CONSTRAINTS
+    func setupConstraints() {
+        
+        iconImageView.snp.updateConstraints {make in
+            make.left.equalToSuperview().offset(24)
```
Comment (drewsdunne) [inline Fitness/Views/Headers/FacilityHoursHeaderView.swift]: Bring up constants.

File: Fitness/Views/GymFacilitiesView.swift (hunk 11 above)
Comment (jhseung) [reply Fitness/Views/GymFacilitiesView.swift]: This is so obvious why did I miss this? 😑 I even consulted. Ok. Yeah that makes so much more sense than this multiple tableview nonsense - sections literally is so much simpler. Wow how did I not even think of that after thinking of so many other ways? Going to fix, and will probably result in significant changes and fixes as well.

File: Fitness/Views/GymFacilitiesView.swift (hunk 10 above)
Comment (jhseung) [reply Fitness/Views/GymFacilitiesView.swift]: Probably not in the near future. It's located in a very wonky place, and given that it's static data (that's also little), it's not a top priority from backend perspective

File: Fitness/Views/GymFacilitiesView.swift (hunk 11 above)
Comment (drewsdunne) [reply Fitness/Views/GymFacilitiesView.swift]: No problem! That's what these reviews are for!

File: Fitness.xcodeproj/project.pbxproj [hunk 15]
```diff
@@ -769,10 +781,10 @@
 				ASSETCATALOG_COMPILER_APPICON_NAME = AppIcon;
 				CODE_SIGN_IDENTITY = "iPhone Developer";
 				CODE_SIGN_STYLE = Automatic;
-				DEVELOPMENT_TEAM = "";
+				DEVELOPMENT_TEAM = A5GB6MS7YX;
 				INFOPLIST_FILE = Fitness/Info.plist;
 				LD_RUNPATH_SEARCH_PATHS = "$(inherited) @executable_path/Frameworks";
-				PRODUCT_BUNDLE_IDENTIFIER = com.cornellappdev.uplift.ios;
+				PRODUCT_BUNDLE_IDENTIFIER = com.cornellappdev.uplift.ioss;
```
Comment (kevinchan159) [inline Fitness.xcodeproj/project.pbxproj]: same here

File: Fitness/Controllers/GymDetailViewController.swift [hunk 16]
```diff
@@ -224,6 +222,20 @@ class GymDetailViewController: UIViewController {
         titleLabel.text = gym.name
         contentView.addSubview(titleLabel)
     }
+    
+    // MARK: - SETUP FACILITIES TABLE
+    func setupFacilities() {
+        if (gym.id == "7045d11329b3645c93556c5aaf44bb21d56934f5") {
+            facilityTableView = GymFacilitiesTableView(facilityName: .helenNewman)
+        } else if (gym.id == "939c7a2c16d2299cc8558475a8007defc414069c" || gym.id == "043c57f3b63411c7a3500c0986fa4b1c8712798c") {
```
Comment (kevinchan159) [inline Fitness/Controllers/GymDetailViewController.swift]: can we make these gym id's `let` constants?

File: Fitness/Views/Headers/FacilityHoursHeaderView.swift [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+    
+    override init(reuseIdentifier: String?) {
+        super.init(reuseIdentifier: reuseIdentifier)
+        
+        facilityNameLabel = UILabel()
+        facilityNameLabel.font = ._16MontserratMedium
+        facilityNameLabel.textColor = .fitnessBlack
+        facilityNameLabel.sizeToFit()
+        facilityNameLabel.textAlignment = .left
+        facilityNameLabel.text = ""
```
Comment (kevinchan159) [inline Fitness/Views/Headers/FacilityHoursHeaderView.swift]: i think its unnecessary to set the text to `""` unless you need it to be that for a specific reason

**Pull request merged**